File: backend/app/schemas/replay.py
"""复盘 Pydantic v2 Schemas（P8.1）。

对照：
- API_SPEC §7 GET /replay/sessions /sessions/{id} /sessions/{id}/snapshots /key-events
- DATA_CONTRACTS §4.12 replay_sessions.summary（P8.1 扩展加 snapshots / key_events
  两个数组字段；前 7 个聚合字段保留契约不变）

体积说明：
- snapshots 按 1Hz × N 任务期录帧，单 session 上限 replay_session_max_frames（默认 1800
  即 30 分钟）。Snapshot 内 blackboard 仅落统计计数，不落全量 entries，避免 MB 级膨胀。
"""
from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.common import Position


# ============== 单帧 / 单事件 ==============


class RobotFrame(BaseModel):
    robot_id: UUID
    code: str
    fsm_state: str
    position: Position | None = None
    battery: float = Field(ge=0.0, le=100.0)
    current_task_id: UUID | None = None


class TaskFrame(BaseModel):
    task_id: UUID
    code: str
    status: str
    progress: float = Field(ge=0.0, le=100.0)
    assigned_robot_ids: list[UUID] = Field(default_factory=list)


class BlackboardFrame(BaseModel):
    """仅统计；避免 entries 全量进 summary 导致 JSONB 爆炸。"""

    total_entries: int = 0
    by_type: dict[str, int] = Field(default_factory=dict)


class Snapshot(BaseModel):
    ts: datetime
    robots: list[RobotFrame] = Field(default_factory=list)
    tasks: list[TaskFrame] = Field(default_factory=list)
    blackboard: BlackboardFrame = Field(default_factory=BlackboardFrame)


KeyEventType = Literal[
    "task_completed",
    "task_failed",
    "task_cancelled",
    "task_reassigned",
    "intervention",
    "alert",
    "auction_completed",
    "recall",
]


class KeyEvent(BaseModel):
    ts: datetime
    type: KeyEventType
    description: str
    related_id: UUID | None = None


# ============== ReplaySession 汇总 ==============


class YoloDetectionsSummary(BaseModel):
    """对应 §4.12 yolo_detections_summary 字面字段。"""

    survivor: int = 0
    fire: int = 0
    smoke: int = 0
    collapsed_building: int = 0


class ReplaySummary(BaseModel):
    """§4.12 原 7 字段 + P8.1 扩展 snapshots / key_events。"""

    total_tasks: int = 0
    completed_tasks: int = 0
    failed_tasks: int = 0
    total_robots_used: int = 0
    total_interventions: int = 0
    total_alerts: int = 0
    yolo_detections_summary: YoloDetectionsSummary = Field(
        default_factory=YoloDetectionsSummary
    )
    snapshots: list[Snapshot] = Field(default_factory=list)
    key_events: list[KeyEvent] = Field(default_factory=list)


class ReplaySessionRead(BaseModel):
    """列表行 + 详情通用：summary 始终返回完整结构。"""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str
    scenario_id: UUID | None = None
    algorithm: str
    started_at: datetime
    ended_at: datetime | None = None
    duration_sec: int | None = None
    completion_rate: float | None = None
    summary: ReplaySummary = Field(default_factory=ReplaySummary)
    created_by: UUID
    created_at: datetime

File: frontend/src/api/replay.ts
import type { AxiosResponse } from 'axios';

import { api } from '@/api/client';
import type { Page } from '@/api/alerts';
import type { Position } from '@/api/robots';

export type ReplayAlgorithm = 'AUCTION_HUNGARIAN' | 'GREEDY' | 'RANDOM' | string;

export interface RobotFrame {
  robot_id: string;
  code: string;
  fsm_state: string;
  position: Position | null;
  battery: number;
  current_task_id: string | null;
}

export interface TaskFrame {
  task_id: string;
  code: string;
  status: string;
  progress: number;
  assigned_robot_ids: string[];
}

export interface Snapshot {
  ts: string;
  robots: RobotFrame[];
  tasks: TaskFrame[];
  blackboard: {
    total_entries: number;
    by_type: Record<string, number>;
  };
}

export interface KeyEvent {
  ts: string;
  type:
    | 'task_completed'
    | 'task_failed'
    | 'task_cancelled'
    | 'task_reassigned'
    | 'intervention'
    | 'alert'
    | 'auction_completed'
    | 'recall';
  description: string;
  related_id: string | null;
}

export interface ReplaySummary {
  total_tasks: number;
  completed_tasks: number;
  failed_tasks: number;
  total_robots_used: number;
  total_interventions: number;
  total_alerts: number;
  yolo_detections_summary: {
    survivor: number;
    fire: number;
    smoke: number;
    collapsed_building: number;
  };
  snapshots: Snapshot[];
  key_events: KeyEvent[];
}

export interface ReplaySessionRead {
  id: string;
  name: string;
  scenario_id: string | null;
  algorithm: ReplayAlgorithm;
  started_at: string;
  ended_at: string | null;
  duration_sec: number | null;
  completion_rate: number | null;
  summary: ReplaySummary;
  created_by: string;
  created_at: string;
}

export interface ReplaySessionListParams {
  start_time?: string;
  end_time?: string;
  algorithm?: string;
  scenario_id?: string;
  page?: number;
  page_size?: number;
}

export async function listReplaySessions(
  params: ReplaySessionListParams = {},
): Promise<Page<ReplaySessionRead>> {
  const r: AxiosResponse<Page<ReplaySessionRead>> = await api.get('/replay/sessions', { params });
  return r.data;
}

export async function getReplaySession(id: string): Promise<ReplaySessionRead> {
  const r: AxiosResponse<ReplaySessionRead> = await api.get(`/replay/sessions/${id}`);
  return r.data;
}

export async function listReplaySnapshots(
  id: string,
  params: { start_time?: string; end_time?: string; interval_sec?: number } = {},
): Promise<Snapshot[]> {
  const r: AxiosResponse<Snapshot[]> = await api.get(`/replay/sessions/${id}/snapshots`, {
    params,
  });
  return r.data;
}

export async function listReplayKeyEvents(
  id: string,
  params: { start_time?: string; end_time?: string } = {},
): Promise<KeyEvent[]> {
  const r: AxiosResponse<KeyEvent[]> = await api.get(`/replay/sessions/${id}/key-events`, {
    params,
  });
  return r.data;
}
